src/Console/ConsoleOutput.php

```diff
@@ -78,25 +81,12 @@ public function logBadMsiErrorMessage(MetricsCalculator $metricsCalculator, floa
         $this->io->error(
             sprintf(
                 self::CI_FLAG_ERROR,
-                'MSI',
+                ($type === TestRunConstraintChecker::MSI_FAILURE ? 'MSI' : 'Covered Code MSI'),
```

Comment:
This a string v. string comparison, `===` is unnecessary here since no type juggling could happen here.

Reply:
`===` is unnecessary but I would argue it's safer to use `===` everywhere and use `==` only when strictly necessary

Reply:
I oppose, since any use of `===` is inherent sign of a *bad* design. There only a small number of cases where one really has to use `===`, all other cases can be avoided by setting explicit return types, or by other means like not using equality comparison in the first place. 

Should we get used to bad design and use `===` to work around it, or should we fight it with all our might? I don't see any another choice here.

Reply:
Woa. I must admit I did not expect that PoV. We might have different opinions about what bad & good design is (and that's fine, it's a subject there will always be disagreement on). I would however argue that using `===` is _safer_. It provides less surprises and risks, so rather than trying to figure out when `==` is sufficient, it is easier to use `===`. There is only two trade-offs that I see:

- Sometimes `==` _is_ necessary, notably with value objects
- It requires one more character to type

Which IMO we can cope with for the benefit brought by it. 

>Should we get used to bad design and use === to work around it, or should we fight bad designs with all our might? I don't see any another choice here.

I would argue the bad design here is not about using `===` or `==`, but about PHP itself. But that doesn't really help since we are in PHP and PHP is not gonna change this feature anytime soon if ever. So I would rather be pragmatic and play it safe, especially if it costs no extra efforts.

Reply:
Sure `===` is safer, but *smells* funny all the same. One thing is when you can't change how `strpos` works, another thing where that's you who are building an API method that may return either `null` or `false` or `0`. These kinds of API are bad is so many ways I wouldn't even try to describe. They are bad for commercial products because every bad API is time and money in the end, spend on fighting it. For open source these things are even more important to avoid since almost nobody is getting paid here.

Good news we are not having to deal with those nasty styles of API, we all good in this regard, so why should we even start to prepare for them?

Reply:
When I see `===` anywhere in the code I immediately start questioning myself: why is this here? What for? Against what is it supposed to guard against? Am I missing something? Why would if be needed? What was that other person thinking?.. And here I am, spending several minutes of my unpaid time on an unimportant yet unobvious peculiarity just because someone decided to play it safe. 

This is not the first time I'm arguing against `===`, so before I asked around, and it turns out almost everyone has to spend time researching every use `===` up to a point where they became desensitized to any use of `===` and just think of them like of regular equals comparison. That's where tide turns for worse because if you ignore `===`, you're missing signals, you're not seeing bad API at the right time. Ignoring `===` makes things worse.

Reply:
I can't help to say but I had a chuckle when reading this, because this is _precisely_ my feeling but when facing `==` instead of `===`.

At the core the issue is this one: do you prefer static typing or not. If yes, you default to strict types and when you omit them, it's for a very specific reason and vice-versa if you prefer dynamic typings. Now we can have fun debating which side is better and why are we debating this in PHP which is at its core a hybrid organic language, but it would be derailing a bit of the original discussion.

So at the core I think it's a question of which side we want to pick for this project. I admit and never hid that I tend to prefer strict types and guards like https://github.com/beberlei/assert or https://github.com/webmozart/assert. It's not pretty to fail with an `InvalidArgumentException`, but I find more efficient to bail out early with an error easy to report which helps to fix the issue fast. But I also worked on several loosely typed code bases so I'm fine and respect the other approach as well. I just think we need to agree on which one to pick here :)

@infection/core thoughts?

Reply:
Personally i'm in the defaul to `===` and only use `==` when absolutely necessary camp, but i can understand the other side as well.

This does mean that infection users may be part of camp `===` or camp `==`. So i'd like to propose to add a `==` => `===` and `!=` => `!==` mutator as well, move all 4 of them to two new profiles, and disable them both by default.

This means we can cater to people who prefer `===` and those who prefer `==` with minimal configuration.

For our code base, i'm with @theofidry on this one.

Reply:
You can't really test `===` in most cases, that's what Infection shows with a low MSI score. If you really care about MSI score, you should consider using `==`, else just move on. All in all I don't think anyone's annoyance by these mutators is valid reason to disable them. We are about technical quality of software, not about ethical, aren't we?

Reply:
As for static typing, I think that for any API is best to stick to specific return/input types as much as possible. This way a user would never ever need to use `===` just because all types are known beforehand.

@theofidry Is that what you assume under static typing?

Reply:
> If you really care about MSI score, you should consider using `==`

I care about MSI in the same way i care about code coverage. I care about it when it helps me understand/track how my test suite is doing. If it becomes lower due to reasons i disagree with, e.g. `==` instead of `===` it becomes a useless metric.

> I don't think anyone's annoyance by these mutators is valid reason to disable them.

When i use a tool, infection or anything else, i want that tool to work for me. I want it to tell me the things i want to know, without too much of a hassle. I completely disagree with changing my code base, simply because my tool tells me i should, especially on opinionated things. (e.g. `==` vs `===`). I'm perfectly fine if those opinionated things align with how the project was already ran/how i want to run the project.

> We are about technical quality of software, not about ethical, aren't we?

Of course, and for mutators like `+` => `-` its obvious what we are testing etc, but `===` => `==` and vice versa, becomes quite opinionated, or ethical if you will.

Reply:
Opinions aren't going well with logical arguments, that's understood. But they are also respectable arguments at times like these, simply because this should be fun and joy, but when your opinions are not respected, it is not anymore.

In this regard, what would you think if we would keep status quo ante? That would mean that I will retract my requests on this PR, and I would not ask anyone else in the future to replace redundant `===`, and I expect that all `===` mutators will be left alone. They're a killer feature for me, just so you would know.

Reply:
>For one because of listed above, for another because PHP is a dynamic language, which is a great power. Best is to embrace and understand these powers, not just blindly follow workarounds, invented by someone else.

But the argument is as valid the other way around: PHP can allow some strictness which is a great power.

>Best is to embrace and understand these powers, not just blindly follow workarounds, invented by someone else. Otherwise put, if you prefer ===, then there should be some reasoning behind this decision.

My reasoning is the following: when I compare `'1'` and `1`, I expect the result to be `false` not `true`. I am not looking for a loose comparison by default, it's not a workaround, this is what I want. Your point is from what I understand whether `===` is _necessary_ to compare it, i.e. if you can safely compare the two values without loose comparison, but IMO even when that's the case it shows the wrong intent. So it's not just a workaround to me: it expresses better the intent I have besides being safe enough to be a sensible default.

>As for static typing, I think that for any API is best to stick to specific return/input types as much as possible. This way a user would never ever need to use === just because all types are known beforehand.

"Strict typing" may be wrongly used my bad. I mean strictness to the code: strict comparison, strict declaration, types whenever possible and in my case I would also add some guards via an assertion library as well. It's not 100% as there is valid cases for duck typing or loose comparisons, but this is rare enough for me to loosen the strictness when necessary.

>All in all I don't think anyone's annoyance by these mutators is valid reason to disable them. We are about technical quality of software, not about ethical, aren't we?

I agree with the message. But mutation testing is expansive so we need to considerate what mutator brings the best value and have a sensible default to have a reasonable running time (i.e. make Infection as fast as possible) and report the least number of false positive. The further one can go the more exigent the profile can be, but if it turns out `===` -> `==` has too many false positives in most code bases I would argue its place is in more exigent profiles, not the default one

Reply:
> when I compare `'1'` and `1`

I see that if you *ever* need to compare these, it's an API problem or just a bug. Bugs should be properly fixed, but not worked around with `===` or else.

Reply:
> Opinions aren't going well with logical arguments, that's understood. But they are also respectable arguments at times like these, simply because this should be fun and joy, but when your opinions are not respected, it is not anymore.

My apologies if i came off harsh or disrespectfull, that wasn't my intention. I simply wanted to state that tools should cater to the users needs, and if there isn't a clear 'better' alternative that it should be configurable.

Reply:
>I see that if you ever need to compare these, it's an API problem or just a bug. Bugs should be properly fixed, but not worked around with === or else.

I agree I usually make sure I do not compare those, but this is PHP: it is neither a strongly typed language neither do we have a compiler to warn us about those. Also mistakes can happen, so I would rather stick to `===` which is safer, is more correct to my intent and has absolutely 0 drawback rather than trying to be more clever than I am and figure out each time if it's really necessary or not.

Reply:
`'0.' == '0.0'` and `==` makes no sense (ever) anymore.

Reply:
That's a nice example where one would actually need to use `===`, but if you need to do it just like than, you're doing it wrong in the first place. It hardly makes any sense to do it just like that be it with `==` or with `===`. Floats must be floats, not strings.

Reply:
Just don't use `==` please. Like, never ever.

Reply:
@Ocramius would you care to explain your position?

Reply:
https://github.com/infection/infection/pull/379#discussion_r198437386

Reply:
Very well, so, where exactly in this PR do you see this type of flawed comparison?

Reply:
I simply exclude these issues upfront, because you just drop more and more time bombs in the codebase, as they become regressions later on. Every `==` removed is an overall betterment that should be welcome (and tbh I don't even understand how this thread got this long by now).

Or you can just install something like `doctrine/coding-standard`, which overall deny the usage of this operator (among multiple flawed PHP features), and stop having these discussions overall. PHPStan strict rules would probably help too.

Reply:
Do you think you could open a PR with these additions, e.g. with PHPStan? Just to end this argument once and for all.

(Actually I'd prefer if PHP-CS-Fixer would do that, but it can't.)

Reply:
Not sure if I can fit it within my schedule, but I'll keep it in mind 👌

Reply:
If you could point me at the right example, that'd be just about same as if you would make a PR.